src/Grpc.Net.Client/Balancer/BalancerAttributes.cs

```diff
@@ -135,10 +159,55 @@ public void Set<TValue>(BalancerAttributesKey<TValue> key, TValue value)
     /// </returns>
     public bool Remove<TValue>(BalancerAttributesKey<TValue> key)
     {
+        ValidateReadOnly();
         return _attributes.Remove(key.Key);
     }
 
-    internal string DebuggerToString()
+    private void ValidateReadOnly()
+    {
+        if (_readOnly)
+        {
+            throw new NotSupportedException("Collection is read-only.");
+        }
+    }
+
+    internal static bool DeepEquals(BalancerAttributes? x, BalancerAttributes? y)
+    {
+        var xValues = x?._attributes;
+        var yValues = y?._attributes;
+
+        if (ReferenceEquals(xValues, yValues))
+        {
+            return true;
+        }
+
+        if (xValues == null || yValues == null)
+        {
+            return false;
+        }
+
+        if (xValues.Count != yValues.Count)
+        {
+            return false;
+        }
+
+        foreach (var kvp in xValues)
+        {
+            if (!yValues.TryGetValue(kvp.Key, out var value))
```

Comment:
Is it interesting to consider the case where the dictionaries have different key comparers?

Reply:
The dictionaries will never have different key comparers.